src/transformers/models/hiera/modeling_hiera.py

```diff
@@ -0,0 +1,1594 @@
+        hidden_states (torch.Tensor): The hidden states tensor of shape [batch_size, num_mask_unit_height*num_mask_unit_width, hidden_size].
+        shape (List[int]): The original shape of the hidden states tensor before windowing.
+        mask_unit_shape (List[int]): The shape of the mask units used for windowing.
+
+    Returns:
+        torch.Tensor: The restored hidden states tensor of shape [batch_size, num_mask_unit_height*mask_unit_height, num_mask_unit_width*mask_unit_width, hidden_size].
+    """
+    num_dims = len(shape)
+    batch_size, hidden_size = hidden_states.shape[0], hidden_states.shape[-1]
+    # From: [batch_size, num_mask_unit_height*num_mask_unit_width, hidden_size]
+    # To: [batch_size, num_mask_unit_height, num_mask_unit_width, mask_unit_height, mask_unit_width, hidden_size]
+    num_mask_units = [s // mu for s, mu in zip(shape, mask_unit_shape)]
+    hidden_states = hidden_states.view(batch_size, *num_mask_units, *mask_unit_shape, hidden_size)
+
+    # From: [batch_size, num_mask_unit_height, num_mask_unit_width, mask_unit_height, mask_unit_width, hidden_size]
+    # To: [batch_size, num_mask_unit_height*mask_unit_height, num_mask_unit_width*mask_unit_width, hidden_size]
+    permute = (
+        [0]
+        + [item for pair in zip(range(1, 1 + num_dims), range(1 + num_dims, 1 + 2 * num_dims)) for item in pair]
+        + [len(hidden_states.shape) - 1]
+    )
+    hidden_states = hidden_states.permute(permute).reshape(batch_size, *shape, hidden_size)
+
+    return hidden_states
+
+
+class HieraEncoder(nn.Module):
+    def __init__(self, config: HieraConfig) -> None:
+        super().__init__()
+        self.config = config
+
+        # stochastic depth decay rule
+        dpr = [x.item() for x in torch.linspace(0, config.drop_path_rate, sum(config.depths))]
+        # query strides rule
+        stage_ends = [sum(config.depths[:i]) - 1 for i in range(1, len(config.depths) + 1)]
+        query_pool_layer = [stage_end + 1 for stage_end in stage_ends[: config.num_query_pool]]
+        query_strides = [
+            math.prod(config.query_stride) if i in query_pool_layer else 1 for i in range(sum(config.depths))
+        ]
+
+        # Transformer blocks
+        self.stages = nn.ModuleList()
+        embed_dim = config.embed_dim
+
+        for idx_stage, depth in enumerate(config.depths):
+            dim_out = int(config.embed_dim * config.embed_dim_multiplier**idx_stage)
+
+            stage = HieraStage(
+                config=config,
+                depth=depth,
+                dim=embed_dim,
+                dim_out=dim_out,
+                num_heads=config.num_heads[idx_stage],
+                drop_path=dpr[sum(config.depths[:idx_stage]) : sum(config.depths[: idx_stage + 1])],
+                query_stride=query_strides[sum(config.depths[:idx_stage]) : sum(config.depths[: idx_stage + 1])],
+                window_size=int(math.prod(config.masked_unit_size) * math.prod(config.query_stride) ** -idx_stage),
```

Comment:
We can have `int(math.prod(config.masked_unit_size) * math.prod(config.query_stride)` as a variable outside the for loop and raise it to `-idx_stage` instead of recalculating it at every iteration

Reply:
https://github.com/huggingface/transformers/pull/30356/commits/7e42e01c4326b095911d5c1e05a1a1c6e279be7a